corehq/apps/es/client.py

```diff
@@ -0,0 +1,994 @@
+        (JSON) and report to datadog.
+        The ``commcare.es.partial_results`` metric counts 1 per Elastic request
+        with any shard failure.
+
+        :param result: Elasticsearch ``search`` or ``scroll`` result object
+        :raises: ``ESShardFailure``, ``ValueError``
+        """
+        if not isinstance(result, dict):
+            raise ValueError(f"invalid Elastic result object: {result}")
+        if result.get("_shards", {}).get("failed"):
+            metrics_counter("commcare.es.partial_results")
+            # Example message:
+            #   "_shards: {"successful": 4, "failed": 1, "total": 5}"
+            shard_info = json.dumps(result["_shards"])
+            raise ESShardFailure(f"_shards: {shard_info}")
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__} index={self.index!r}, type={self.type!r}>"
+
+
+class BulkActionItem:
+    """A "wrapper" for documents to be processed via Elasticsearch's Bulk API.
+    Collections of these objects can be passed to an ElasticDocumentAdapter's
+    bulk() method for processing.
+
+    Instances of this class are meant to be acquired via one of the factory
+    methods rather than instantiating directly (via ``__init__()``).
+    """
+
+    OpType = Enum("OpType", "index delete")
+
+    def __init__(self, op_type, doc=None, doc_id=None):
+        if not (isinstance(op_type, self.OpType) and op_type in self.OpType):
+            raise ValueError(f"invalid operations type: {op_type!r}")
+        if doc is None and doc_id is None:
+            raise ValueError("at least one of 'doc' or 'doc_id' are required")
+        elif not (doc is None or doc_id is None):
+            raise ValueError("'doc' and 'doc_id' are mutually exclusive")
+        elif doc is None and op_type is not self.OpType.delete:
+            raise ValueError("'doc_id' can only be used for delete operations")
+        self.doc = doc
+        self.doc_id = doc_id
+        self.op_type = op_type
+
+    @classmethod
+    def delete(cls, doc):
+        """Factory method for a document delete action"""
+        return cls(cls.OpType.delete, doc=doc)
+
+    @classmethod
+    def delete_id(cls, doc_id):
+        """Factory method for a document delete action providing only the ID"""
+        return cls(cls.OpType.delete, doc_id=doc_id)
+
+    @classmethod
+    def index(cls, doc):
+        """Factory method for a document index action"""
+        return cls(cls.OpType.index, doc=doc)
+
+    @property
+    def is_delete(self):
+        """Return ``True`` if this action is for a document delete operation."""
+        return self.op_type is self.OpType.delete
+
+    @property
+    def is_index(self):
+        """Return ``True`` if this action is for a document index operation."""
+        return self.op_type is self.OpType.index
+
+    def __repr__(self):
+        human_op = getattr(self.op_type, "name", "???")  # avoid AttributeError
```

Comment:
nit: In what situation does `self.op_type.name` not exist?
```suggestion
        human_op = self.op_type.name
```